### PR TITLE
[Secure JSON RPC] Implement a lightweight JSON RPC client for secure components.

### PR DESCRIPTION
**Pull request by JoshLind**
## Motivation

In this PR, we create a lightweight JSON RPC client to allow security-critical components (e.g., the key manager) to communicate with the Libra blockchain using the JSON RPC interface offered by full nodes. In comparison to the already existing JSON RPC client, the lightweight client:

1) Uses ureq instead of reqwest and avoids async dependencies.
2) Removes the use of transaction-builder.
3) Offers a small(-ish) crate that separates the client code from the server, allowing fewer dependencies in the TCB.
4) Only exposes the functionality required by the key manager: (i) transaction submission; and (ii) retrieving account state. More functionality may be added in the future..

We note that the tests written here for the lightweight client require dependencies on the original json-rpc crate (to avoid duplicating several large pieces of test code/functionality). However, such dependencies are only required for testing and are not included in the actual client binary/code. If it is deemed necessary, we can work on removing these test dependencies in a future PR, but this will likely require duplicating a lot of code.

Also, if it helps, 'cargo geiger' shows the lightweight client crate has less (5x?) unsafe code than the original JSON RPC crate (including testing code). However: (i) 'cargo geiger' crashes half way through analyzing the original JSON RPC crate, so I'm only comparing a subset of the dependencies of the original crate..; and (ii) the original JSON RPC crate also contains the server, so it's not a fair comparison to make until someone manages to separate the client/server code in the original crate. I would be really interested in comparing the two clients... However, this is currently out of the scope for now.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

All tests pass locally (including the newly added ones.)

## Related PRs

None.

### CODE REVIEW

File: secure/json-rpc/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use hex::FromHexError;
+use std::io;
+use thiserror::Error;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum Error {
+    #[error("JSON RPC call {0} returned a custom internal error: {0}")]
+    InternalRPCError(String, String),
```
Comment (davidiw) [inline secure/json-rpc/src/error.rs:11]: do we benefit by leaking which call made the error?

File: secure/json-rpc/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    #[error("Data does not exist. Missing data: {0}")]
+    MissingData(String),
+    #[error("JSON RPC call {0} failed. Response: {0}")]
+    RPCFailure(String, String),
```
Comment (davidiw) [inline secure/json-rpc/src/error.rs:15]: Same here, since this adds context that should already be there. We would want to add context outside this error for more information afaiu.

File: secure/json-rpc/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    #[error("Serialization error: {0}")]
+    SerializationError(String),
+    #[error("Unknown error: {0}")]
+    UnknownError(String),
+}
+
+impl From<anyhow::Error> for Error {
```
Comment (davidiw) [inline secure/json-rpc/src/error.rs:22]: Please do not introduce anyhow into this crate unless it is blocked by a feature. I do not want anyhow ... if at all possible .. in our TCB.

Comment (davidiw) [reply secure/json-rpc/src/error.rs:22]: Motivations:
* Lots of unsafe code
* Implies soft definition of error types

File: secure/json-rpc/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+#[cfg(test)]
+mod tests;
+
+/// Various constants for the JSON RPC client implementation
+const JSON_RPC_VERSION: &str = "2.0";
+const REQUEST_TIMEOUT: u64 = 10_000;
+
+/// This defines a simple and generic interface by which to interact with the JSON RPC service
+/// offered by each Libra Full Node.
+pub trait JsonRpcClient {
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:33]: why do we need this?

File: secure/json-rpc/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    /// using the JSON RCP API.
+    fn get_account_state(
+        &self,
+        account: AccountAddress,
+        version: u64,
+    ) -> Result<AccountState, Error>;
+}
+
+/// Provides a lightweight JsonRpcClient implementation.
+struct LightweightJsonClient {
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:48]: I don't think we need the additional naming here... it is already in a different namespace... Simply JsonRpcClient would do.

File: secure/json-rpc/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            Value::String(account.to_string()),
+            Value::Number(version.into()),
+            Value::Number(ledger_version.into()),
+        ];
+        Ok(self.execute_request(method, params))
+    }
+
+    /// Parses the response returned by the submit() JSON RPC request, and ensures the request
+    /// was successful.
+    fn process_submit_response(&self, response: Response) -> Result<(), Error> {
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:100]: is there a reason to not have this inline where it is being used only once? It makes it easier to review the code.

File: secure/json-rpc/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        } else {
+            let _submit_response = serde_json::from_str::<SubmitTransactionResponse>(response)?;
+            Ok(())
+        }
+    }
+
+    /// Parses the response returned by the get_account_state_with_proof() JSON RPC request. For
+    /// this, we verify that the request was successful, and return the corresponding AccountState
+    /// provided by the JSON RPC server; fail otherwise.
+    fn process_account_with_proof_response(
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:116]: again, why are we separating these?

File: secure/json-rpc/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        let request_id = 0;
+        ureq::post(&self.host)
+            .timeout_connect(REQUEST_TIMEOUT)
+            .send_json(
+                json!({"jsonrpc": JSON_RPC_VERSION, "method": method, "params": params, "id": request_id}),
+            )
+    }
+
+    // Sends a submit() request to the JSON RPC server using the given transaction.
+    fn submit(&self, signed_transaction: SignedTransaction) -> Result<Response, Error> {
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:73]: submit_transaction?

File: secure/json-rpc/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    #[cfg(test)]
+    pub fn new(host: &str, port: u16) -> Self {
+        let host = format!("http://{}:{}", host, port);
+        Self { host }
+    }
+
+    // Executes the specified request method using the given parameters by contacting the JSON RPC
+    // server.
+    fn execute_request(&self, method: String, params: Vec<Value>) -> Response {
+        let request_id = 0;
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:64]: if this is always 0, why specify it?

File: secure/json-rpc/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    }
+
+    fn get_account_state(
+        &self,
+        account: AccountAddress,
+        version: u64,
+    ) -> Result<AccountState, Error> {
+        let response = self.get_account_state_with_proof(account, version, version)?;
+        match response.status() {
+            200 => self.process_account_with_proof_response(response),
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:162]: let's just make everything in line?

File: secure/json-rpc/src/error.rs (hunk 2 above)
Comment (JoshLind) [reply secure/json-rpc/src/error.rs:15]: See above.

File: secure/json-rpc/src/error.rs (hunk 3 above)
Comment (JoshLind) [reply secure/json-rpc/src/error.rs:22]: I was hoping to remove it entirely, but it is required for testing (e.g., to implement the DbReader for the mock database). It is also required if we want to handle anyhow errors automatically (e.g., using the '?' operator in Rust). However, I've managed to remove it from non-test dependencies by checking and handling each anyhow error manually now.

File: secure/json-rpc/src/lib.rs (hunk 4 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:33]: I was hoping to abstract away from the implementation, allowing us to switch between the lightweight client and regular client easily, however, this may be a premature optimization, so I've removed the trait for now.

File: secure/json-rpc/src/lib.rs (hunk 7 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:116]: See above.

File: secure/json-rpc/src/lib.rs (hunk 8 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:73]: I was explicitly using the request method names to label these functions to make it obvious (e.g., submit and get_account_state_with_proof), however, I've changed this to submit_transaction as you suggest. I don't have super strong feelings.

File: secure/json-rpc/src/lib.rs (hunk 10 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:162]: See above.

File: secure/json-rpc/src/error.rs (hunk 1 above)
Comment (JoshLind) [reply secure/json-rpc/src/error.rs:11]: I was wanting to make the error message as helpful as possible, i.e., without needing to look at the backtrace, but I guess as long as the backtrace is there, that's fine too!

File: secure/json-rpc/src/lib.rs (hunk 6 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:100]: As discussed offline, this was more a readability thing (as some of the functions are pretty long/nested otherwise), but I guess that seems to be a result of the way Rust works sometimes.. I've inlined the functions.. let me know what you think. Personally I think the 'get_account_state' function is a little long, but it's not super bad!

File: secure/json-rpc/src/tests/client.rs [hunk 11]
```diff
@@ -0,0 +1,209 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{tests::mock_libra_db::MockLibraDB, JsonRpcClient, LightweightJsonClient};
+use futures::{channel::mpsc::channel, StreamExt};
+use libra_config::utils;
+use libra_crypto::{ed25519::Ed25519PrivateKey, HashValue, PrivateKey, Uniform};
+use libra_json_rpc::bootstrap;
```
[the remark this answers is not in the document]
Comment (JoshLind) [reply secure/json-rpc/src/tests/client.rs:8]: Generally, I agree with you. However, in this case, it's seems like someone has already explicitly exported functions from the runtime file in the libra_json_rpc package to avoid making the file public (e.g., see the 'bootstrap_from_config' export in the 'lib.rs' file). If you want to make the entire thing public we can do that, but I wanted to keep the original semantic structure (to avoid issues I may not be aware about...)

File: secure/json-rpc/src/tests/mock_libra_db.rs [hunk 12]
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) The Libra Core Contributors
```
[the remark this answers is not in the document]
Comment (JoshLind) [reply secure/json-rpc/src/tests/mock_libra_db.rs:1]: As discussed offline this was a result of several things: (i) the current mock database not supporting test functionality for what we need; (ii) the current mock database already being pretty complicated, thus pulling in a lot of dependencies we don't necessarily need; and (iii) running into a weird edge case where the compiler can't find the exported mock database (despite my IDE happily compiling everything)..  In the spirit of 'move fast' I decided to just write a small mock containing only what we need so I could make this a reality, but I'll spend a bit of time now trying to see if I can work around the issue that was blocking me..

**Comment by JoshLind** [comment]
Thanks, David! As we discussed offline, I've also merged everything into a single file ('lib.rs'). It's a little long, but it encapsulates everything quite nicely. 😄

File: secure/json-rpc/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,595 @@
...
+}
+
+/// Provides a lightweight JsonRpcClient implementation.
+struct JsonRpcClient {
+    host: String,
+}
+
+// TODO(joshlind): remove this flag once the client is integrated into key manager. Otherwise,
+// rust complains this is never used...
+#[allow(dead_code)]
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:73]: the issue is line 67, you need pub struct

File: secure/json-rpc/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+impl JsonRpcClient {
+    pub fn new(host: &str, port: u16) -> Self {
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:75]: recommend you actually pass in a URL since it could be https or http, granted for our use case it is irrelevant.

File: secure/json-rpc/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,595 @@
...
+        let host = format!("http://{}:{}", host, port);
+        Self { host }
+    }
+
+    /// Submits a signed transaction to the Libra blockchain via the JSON RPC API.
+    pub fn submit_signed_transaction(
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:81]: I noticed today that the other API just has a "submit" function, :/ I like this more though

File: secure/json-rpc/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+    // server.
+    fn execute_request(&self, method: String, params: Vec<Value>) -> Response {
+        ureq::post(&self.host)
+            .timeout_connect(REQUEST_TIMEOUT)
+            .send_json(
+                json!({"jsonrpc": JSON_RPC_VERSION, "method": method, "params": params, "id": 0}),
+            )
+    }
+
+    // Sends a submit() request to the JSON RPC server using the given transaction.
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:152]: why isn't this inline?

File: secure/json-rpc/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,595 @@
...
+        let method = "submit".to_string();
+        let params = vec![Value::String(hex::encode(lcs::to_bytes(
+            &signed_transaction,
+        )?))];
+        Ok(self.execute_request(method, params))
+    }
+
+    // Sends a get_account_state_with_proof() request to the JSON RPC server for the specified
+    // account, version height and ledger_version height (for the proof).
+    fn get_account_state_with_proof(
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:163]: why isn't this inline?

File: secure/json-rpc/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,595 @@
...
+            .timeout_connect(REQUEST_TIMEOUT)
+            .send_json(
+                json!({"jsonrpc": JSON_RPC_VERSION, "method": method, "params": params, "id": 0}),
+            )
+    }
+
+    // Sends a submit() request to the JSON RPC server using the given transaction.
+    fn submit_transaction(&self, signed_transaction: SignedTransaction) -> Result<Response, Error> {
+        let method = "submit".to_string();
+        let params = vec![Value::String(hex::encode(lcs::to_bytes(
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:155]: I wonder why they went with hex instead of base64....

File: secure/json-rpc/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,595 @@
...
+/// A custom error returned by the JSON RPC server for API calls that fail internally (e.g.,
+/// an internal error during execution on the server side).
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct JsonRpcError {
+    code: i16,
+    message: String,
+    data: Option<Value>,
+}
+
+/// Represents a vector of bytes using hex encoding.
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:249]: should we consider making this an LCS data structure Bytes<T> where T is Serialize / Deserialize? I am not sure if that would save you lines of code, but this is effectively an Hex(LCS) wrapper around types T.

File: secure/json-rpc/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,595 @@
...
+}
+
+impl From<&Vec<u8>> for Bytes {
+    fn from(bytes: &Vec<u8>) -> Self {
+        Self(hex::encode(bytes))
+    }
+}
+
+/// Testing infrastructure for the lightweight JSON client above. Hooks into the original JSON RPC
+/// infrastructure.
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:272]: no need for doc block or really even the comment, it is intuitive.

File: secure/json-rpc/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,595 @@
...
+#[cfg(test)]
+mod test {
+    use crate::JsonRpcClient;
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:275]: always recommend using super::* first and then pulling in additional dependencies that aren't in super::*
this is the only time I /ever/ recommend using * :P

File: secure/json-rpc/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,595 @@
...
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_submit_transaction_failure() {
+        let mock_db = create_empty_mock_db();
+        // When creating the JSON RPC server, specify 'mock_validator=false' to prevent a vm validator
+        // being passed to the server. This will cause any transaction submission requests to the JSON
+        // RPC server to fail, thus causing the server to return an error.
+        let (client, _server) = create_client_and_server(mock_db, false);
```
Comment (davidiw) [inline secure/json-rpc/src/lib.rs:327]: if this is the only case you pass a vm_validator, why not just mutate the signed_transaction to be a bad signed transaction? that way you don't need this modified vm_validator?

File: secure/json-rpc/src/lib.rs (hunk 18 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:155]: Yeah.. I saw the discussion you started yesterday!

File: secure/json-rpc/src/lib.rs (hunk 22 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:327]: I did originally try this, but (iirc) even with the badly signed transaction, the call still fails without passing in a vm_validator.. Going to leave it as-is for now!

File: secure/json-rpc/src/lib.rs (hunk 16 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:152]: Personally, I think it's cleaner as a separate method (otherwise the calling functions become excessively long)

File: secure/json-rpc/src/lib.rs (hunk 17 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:163]: See above.

File: secure/json-rpc/src/lib.rs (hunk 19 above)
Comment (JoshLind) [reply secure/json-rpc/src/lib.rs:249]: We could do that in the future, but seeing as though it won't save space, let's just keep it simple for now. Can always revisit it!

**Comment by JoshLind** [comment]
Thanks, David! @bors-libra r=davidiw
